### PR TITLE
issue #4888: show user's vote in poll

### PR DESCRIPTION
**Pull request by netagonen**
Hi! 

Just wanted to let you know that we didn't fix the following checks for the reasons stated below them:

```
app/assets/javascripts/app/views/poll_view.js:36 W: Identifier 'original_post_link' is not in camel case.
app/assets/javascripts/app/views/poll_view.js:37 W: Identifier 'poll_answers' is not in camel case.
```
Here we are following exactly the code that is already there and it seems like it is properly formatted.

```
app/models/poll.rb:28 W: Use the new Ruby 1.9 hash syntax.
app/models/poll.rb:28 W: Redundant `self` detected.
```
These are not changes that we have made.

CC @Rete2

### CODE REVIEW

**Comment by svbergerem** [comment]
Please fix (some of) the code style errors. We still have a lot of old code that doesn't follow our styleguides and we agreed on changing the code to follow the styleguides whenever we touch it.

When fixing the camel case issues you would also have to change the identifiers in the poll template. If you don't want to do that, then I'd say it's okay to leave the code like it is.

The ruby code style issues in `app/models/poll.rb` should be fixed.

You will also have to add a fix for the dark theme (in `vim app/assets/stylesheets/color_themes/_color_theme_override_dark.scss`).

poll dark theme

**Comment by netagonen** [comment]
> When fixing the camel case issues you would also have to change the identifiers in the poll template. If you don't want to do that, then I'd say it's okay to leave the code like it is.

We changed the camel case also in the template.

> The ruby code style issues in app/models/poll.rb should be fixed.

We changed lines 24-28 the the new hash syntax and not only line 28. 
Regarding the `Redundant `self``: what does that mean?

**Comment by SuperTux88** [comment]
> Regarding the `Redundant 'self'`: what does that mean?

That means, that you don't need `self` there, you can simply remove it.

**Comment by netagonen** [comment]
Hi, we fixed the style errors. BUT we are still working on adding a label next to the progress-bar, so it's not ready for a full review.

**Comment by SuperTux88** [comment]
By the way: You can add `[WIP]` (work in progress) to a PR in the title, so it's visible to us, that you are still working on it (but you want to ask a question or need some feedback about something). I added that now, you can remove it (and write a comment, because title changes don't send notifications) when you think you are finished and it's ready for a full review (only the person who opened the PR can edit the title).

**Comment by netagonen** [comment]
We are now ready for a full review :)
We only got this error that we are not sure how to fix, can you give us a hint how to?
`.eslintrc: F: Parsing error: Unexpected token }`

File: app/assets/javascripts/app/views/poll_view.js [hunk 1]
```diff
@@ -21,11 +21,21 @@ app.views.Poll = app.views.Base.extend({
     var originalPostLink = isReshare && this.model.get('root') ?
       '<a href="/posts/' + this.model.get('root').id + '" class="root_post_link">' + Diaspora.I18n.t('poll.original_post') + '</a>' :
       '';
+    var answerGiven = this.model.get("poll_participation_answer_id");
+    var currentUser = defaultPresenter.current_user;
+
+    if (defaultPresenter.poll && defaultPresenter.poll.poll_answers) {
+      var pollAnswers = defaultPresenter.poll.poll_answers;
+      pollAnswers = _.map(pollAnswers, function(answer) {
+        return _.extend(answer, {isCurrentUserVote: answer.id === answerGiven, currentUser});
```
Comment (svbergerem) [inline app/assets/javascripts/app/views/poll_view.js]: This line is causing the syntax error. It should be `currentUser: currentUser`.

Comment (netagonen) [inline app/assets/javascripts/app/views/poll_view.js]: Thanks! fixed! now it's completely ready

File: app/assets/javascripts/app/views/poll_view.js [hunk 2]
```diff
@@ -21,11 +21,21 @@ app.views.Poll = app.views.Base.extend({
...
+        return _.extend(answer, {isCurrentUserVote: answer.id === answerGiven, currentUser: currentUser});
+      });
+    }
 
     return _.extend(defaultPresenter, {
       show_form: showForm,
       is_reshare: isReshare,
-      original_post_link: originalPostLink
+      originalPostLink: originalPostLink,
+      pollAnswers: pollAnswers
```
Comment (svbergerem) [inline app/assets/javascripts/app/views/poll_view.js]: You don't use it in the poll template, so you can just remove it.

File: app/assets/javascripts/app/views/poll_view.js [hunk 3]
```diff
@@ -21,11 +21,21 @@ app.views.Poll = app.views.Base.extend({
...
       '<a href="/posts/' + this.model.get('root').id + '" class="root_post_link">' + Diaspora.I18n.t('poll.original_post') + '</a>' :
       '';
+    var answerGiven = this.model.get("poll_participation_answer_id");
+    var currentUser = defaultPresenter.current_user;
+
+    if (defaultPresenter.poll && defaultPresenter.poll.poll_answers) {
+      var pollAnswers = defaultPresenter.poll.poll_answers;
+      pollAnswers = _.map(pollAnswers, function(answer) {
+        return _.extend(answer, {isCurrentUserVote: answer.id === answerGiven, currentUser: currentUser});
+      });
```
Comment (svbergerem) [inline app/assets/javascripts/app/views/poll_view.js]: The only purpose of these four lines is changing `defaultPresenter.poll.poll_answers` and I'd like to make that more explicit. Maybe something like this:

```JavaScript
defaultPresenter.poll.poll_answers.forEach(function (answer) {
  // set answer.isCurrentUserVote and answer.currentUser
});
```

File: app/assets/templates/poll_tpl.jst.hbs [hunk 4]
```diff
@@ -13,13 +13,26 @@
...
+            {{#if isCurrentUserVote}}
+              <label>
+                {{answer}}
+                  {{#linkToAuthor currentUser}}
+                    {{{personImage this 'small' 'micro'}}}
+                  {{/linkToAuthor}}
+              </label>
+            {{else}}
+              <label>{{answer}}</label>
+            {{/if}}
```
Comment (svbergerem) [inline app/assets/templates/poll_tpl.jst.hbs]: The indentation here is a bit weird. `{{#if isCurrentUserVote}}` should be indented like `<input type="radio" name="vote" value="{{id}}"/>` and both `{{answer}}` and `{{#linkToAuthor currentUser}}` like `<label>`.

File: app/assets/templates/poll_tpl.jst.hbs [hunk 5]
```diff
@@ -13,13 +13,26 @@
       {{#poll.poll_answers}}
         <div class="result-row">
           <input type="radio" name="vote" value="{{id}}"/>
-          <label>{{answer}}</label>
+            {{#if isCurrentUserVote}}
+              <label>
+                {{answer}}
+                  {{#linkToAuthor currentUser}}
+                    {{{personImage this 'small' 'micro'}}}
+                  {{/linkToAuthor}}
```
Comment (svbergerem) [inline app/assets/templates/poll_tpl.jst.hbs]: I'm not sure if we should add a link to our own profile or just a simple avatar without the link.

File: spec/javascripts/app/views/poll_view_spec.js [hunk 6]
```diff
@@ -75,4 +75,32 @@ describe("app.views.Poll", function(){
...
+      this.view.model.set("poll_participation_answer_id", answer.id);
+      expect(this.view.$(".poll_progress_bar.users-vote").length).toBe(1);
+    });
+
+    it("doesn't add 'users-vote' class to progress bar of the options the user didn't vote for", function() {
+      var answer1 = this.view.poll.poll_answers[0];
+      var answer2 = this.view.poll.poll_answers[1];
+      this.view.model.set("poll_participation_answer_id", answer1.id);
+      var progressBarAnswer2 = ".poll_progress_bar[data-answerid='" + answer2.id + "']";
+      expect(this.view.$(progressBarAnswer2).hasClass("users-vote")).toBe(false);
```
Comment (svbergerem) [inline spec/javascripts/app/views/poll_view_spec.js]: Using a variable for the selector looks unnecessary to me. You could replace `progressBarAnswer2` by the value set before and still follow the maximum line length rule. (120 chars)

File: spec/javascripts/app/views/poll_view_spec.js [hunk 7]
```diff
@@ -75,4 +75,32 @@ describe("app.views.Poll", function(){
...
+      var answer = this.view.poll.poll_answers[0];
+      this.view.model.set("poll_participation_answer_id", answer.id);
+      expect(this.view.$(".micro.avatar.img-responsive.center-block").length).toBe(1);
+    });
+
+    it("shows the 'brand-primary' color on the option the user voted for", function() {
+      var answer = this.view.poll.poll_answers[0];
+      this.view.model.set("poll_participation_answer_id", answer.id);
+      expect(this.view.$(".poll_progress_bar.users-vote").css("background-color")).toBe("rgb(0, 136, 230)");
+    });
```
Comment (svbergerem) [inline spec/javascripts/app/views/poll_view_spec.js]: You already tested before that the `users-vote` class is added, so here you only check the CSS rule for `.users-vote`. I'd say we can remove this test, since it should only test things our JavaScript code does. @denschub Any other opinion?

**Review by svbergerem — CHANGES_REQUESTED** [review]
These are mostly small remarks. The new color for the progress bars looks good, but there is something about the avatar next to the label that I don't like. Unfortunately I'm currently unable to describe the reason for that. I tried to add some margin to the left of the avatar, but I'm still not satisfied. Maybe someone else has a good idea? I'll add a screenshot of the current solution below.

**Comment by svbergerem** [comment]
Here is a screenshot of the current design:

poll view avatar

**Comment by SuperTux88** [comment]
Hmm, I see what you mean, the avatar doesn't really fit there. While I somehow like it, it also looks strange. 

Also: When a user doesn't have an avatar it would simply display the default fallback avatar, which isn't clear then. So maybe just a text label with "your vote" or something like this is better?

**Comment by Rete2** [comment]
We put a text label next to the answer now. For some reason I can´t post a screenshot here right now. I will try that later.
But one question about translations:
We added a translation in `config/locales/javascript/javascript.en.yml` and use it then in the template for the text in the label. Is this the right thing to do or do we need to add it in a different file?

**Comment by SuperTux88** [comment]
> We added a translation in `config/locales/javascript/javascript.en.yml` and use it then in the template for the text in the label. Is this the right thing to do or do we need to add it in a different file?

Yes, that's the correct file.

File: spec/javascripts/app/views/poll_view_spec.js (hunk 7 above)
Comment (netagonen) [inline spec/javascripts/app/views/poll_view_spec.js]: We removed the test

**Comment by Rete2** [comment]
We´re still working on your comments, so no need to look at the latest commit yet.

**Comment by netagonen** [comment]
We made the corrections :) so it's ready on our part

**Comment by goobertron** [comment]
Sorry, I'm coming late to these latest changes.

I think the avatar is unnecessary: the blue colour of the bar indicates which option the user has voted for; and, as it is merely a reminder of which option they themselves voted for, I don't think it's necessary to reinforce this with their own avatar or a text label.

The avatar also doesn't really fit well in amongst the bars. It it a bit of visual 'noise'. While I haven't seen a screenshot of the text label there, I think this would have the same effect.

I think it's best to keep the design as simple as possible by just altering the bar colour.

**Comment by netagonen** [comment]
At the moment we switched the avatar to a label. Originally we wanted this extra indication for the user's vote because a change of the progress bar color could be interpreted also as "the most popular vote" when it is actually the most popular vote. But if the majority here thinks it's unnecessary, we can drop this change.

**Comment by SuperTux88** [comment]
Here is a screenshot:


I think the label is OK (better than the avatar). I think it's cool to show what the color means. When the colored bar is also the winning one, I think it's maybe confusing if the color is because it's winning or because it's your vote. So I think a indicator would be nice. An alternative would be to add a tooltip when hovering the colored bar and show "your vote" there.

File: config/locales/javascript/javascript.en.yml [hunk 8]
```diff
@@ -344,3 +344,4 @@ en:
         other: "<%=count%> votes"
       show_result: "Show result"
       close_result: "Hide result"
+      vote_label: "Your vote"
```
Comment (SuperTux88) [inline config/locales/javascript/javascript.en.yml]: I think `your_vote` would be a better key than `vote_label`.

File: app/assets/javascripts/app/views/poll_view.js [hunk 9]
```diff
@@ -21,11 +21,19 @@ app.views.Poll = app.views.Base.extend({
     var originalPostLink = isReshare && this.model.get('root') ?
       '<a href="/posts/' + this.model.get('root').id + '" class="root_post_link">' + Diaspora.I18n.t('poll.original_post') + '</a>' :
       '';
+    var answerGiven = this.model.get("poll_participation_answer_id");
+    var currentUser = defaultPresenter.current_user;
+
+    if (defaultPresenter.poll && defaultPresenter.poll.poll_answers) {
+      defaultPresenter.poll.poll_answers.forEach(function(answer) {
+        return _.extend(answer, {isCurrentUserVote: answer.id === answerGiven, currentUser: currentUser});
```
Comment (svbergerem) [inline app/assets/javascripts/app/views/poll_view.js]: You don't need the `return` here. Just write `_.extend(...);`.

File: app/assets/templates/poll_tpl.jst.hbs [hunk 10]
```diff
@@ -13,15 +13,26 @@
...
           <input type="radio" name="vote" value="{{id}}"/>
-          <label>{{answer}}</label>
-          <div class="poll-result pull-right">
-            <span class="percentage"></span>
-            ({{t "poll.answer_count" count=vote_count}})
-          </div>
-          <div class="poll_progress_bar_wrapper progress">
+          {{#if isCurrentUserVote}}
+            <label>
+            {{answer}}
```
Comment (svbergerem) [inline app/assets/templates/poll_tpl.jst.hbs]: I'm sorry if I made a mistake in my comment before (I'm unable to check right now) but here are two whitespaces missing for indentation.

**Comment by Flaburgan** [comment]
There is some work to do to deal with long labels: https://github.com/diaspora/diaspora/issues/7573 this is not part of this PR, but if we add a label "Your vote" at the end, it will be even more common.

File: config/locales/javascript/javascript.en.yml (hunk 8 above)
Comment (svbergerem) [inline config/locales/javascript/javascript.en.yml]: I agree.

**Review by svbergerem — CHANGES_REQUESTED** [review]

**Comment by goobertron** [comment]
Thanks for that screenshot. The label looks a lot better than I imagined; I thought it would be the user's name (which can be very long).

If everyone else wants a label, I'm happy for this to be used.

File: app/assets/templates/poll_tpl.jst.hbs [hunk 11]
```diff
@@ -13,15 +13,26 @@
...
+        </div>
+        <div class="poll_progress_bar_wrapper progress">
+          {{#if isCurrentUserVote}}
+            <div class="poll_progress_bar bar users-vote" data-answerid="{{id}}" style="height: 100%"></div>
+          {{else}}
             <div class="poll_progress_bar bar" data-answerid="{{id}}" style="height: 100%"></div>
-          </div>
+          {{/if}}
         </div>
+      </div>
```
Comment (svbergerem) [inline app/assets/templates/poll_tpl.jst.hbs]: Mhh, it looks like the indentation in this whole block needs some work.

**Comment by svbergerem** [comment]
Let's start with the label. We can still change it in future commits if we think that there is a better solution, but for now I think it's ok.

File: app/assets/javascripts/app/views/poll_view.js [hunk 12]
```diff
@@ -21,11 +21,19 @@ app.views.Poll = app.views.Base.extend({
     var originalPostLink = isReshare && this.model.get('root') ?
       '<a href="/posts/' + this.model.get('root').id + '" class="root_post_link">' + Diaspora.I18n.t('poll.original_post') + '</a>' :
       '';
+    var answerGiven = this.model.get("poll_participation_answer_id");
+    var currentUser = defaultPresenter.current_user;
```
Comment (svbergerem) [inline app/assets/javascripts/app/views/poll_view.js]: You don't use `currentUser` anymore in the template, so you can remove it here.

File: app/assets/javascripts/app/views/poll_view.js [hunk 13]
```diff
@@ -21,11 +21,19 @@ app.views.Poll = app.views.Base.extend({
...
+
+    if (defaultPresenter.poll && defaultPresenter.poll.poll_answers) {
+      defaultPresenter.poll.poll_answers.forEach(function(answer) {
+        _.extend(answer, {isCurrentUserVote: answer.id === answerGiven, currentUser: currentUser});
```
Comment (svbergerem) [inline app/assets/javascripts/app/views/poll_view.js]: And you can remove `currentUser` here as well.

File: app/assets/stylesheets/poll.scss [hunk 14]
```diff
@@ -46,6 +50,10 @@
     font-weight: normal;
     margin-bottom: 5px;
     vertical-align: middle;
+
+    .center-block {
+      display: inline;
+    }
```
Comment (svbergerem) [inline app/assets/stylesheets/poll.scss]: I think this rule is not needed anymore.

**Review by svbergerem — CHANGES_REQUESTED** [review]
Just some cleanup after the latest changes. Everything else in the front-end looks good to me.

**Review by svbergerem — APPROVED** [review]
The front-end code looks good to me. 👍

File: app/presenters/post_presenter.rb [hunk 15]
```diff
@@ -47,6 +47,10 @@ def page_title
     post_page_title @post
   end
 
+  def poll_participation_answer_id
+    @post.poll.participation_answer(current_user).poll_answer_id if already_participated_in_poll
```
Comment (SuperTux88) [inline app/presenters/post_presenter.rb]: You can combine this with `already_participated_in_poll`, so this line looks like this:
```ruby
@post.poll&.participation_answer(current_user)&.poll_answer_id if user_signed_in?
```
and then also use `poll_participation_answer_id` to decide "if already participated" in the `poll_view.js`.

Then you can remove `already_participated_in_poll` from `post_presenter.rb` and remove `already_participated?` from `poll.rb`.

This has two reasons:
1. less code
2. less database queries

with the old way you queried the participation 3 times from the database:
1. for the `if already_participated_in_poll` in this line
2. for the `poll.participation_answer(current_user)` here
3. again `already_participated_in_poll` to add it to `non_directly_retrieved_attributes`.

With the new line you do only one query, because you can decide "if already participated" if this method returns nil or the id.

File: spec/presenters/post_presenter_spec.rb [hunk 16]
```diff
@@ -135,6 +135,14 @@
...
       expect(presenter.as_json).to be_a(Hash)
     end
+
+    it "returns the answer id of the current user's poll participation" do
+      presenter = PostPresenter.new(status_message_with_poll, alice)
+      poll_answer = status_message_with_poll.poll.poll_answers.first
+      poll_participation = status_message_with_poll.poll.poll_participations
+      poll_participation = poll_participation.create(poll_answer: poll_answer, author: alice.person)
+      expect(presenter.poll_participation_answer_id).to eql(poll_participation.poll_answer_id)
+    end
```
Comment (SuperTux88) [inline spec/presenters/post_presenter_spec.rb]: When you do the change in `post_presenter_spec.rb` this would also need a negative test to test it returns nil when the user didn't participate yet.

**Review by SuperTux88 — CHANGES_REQUESTED** [review]

File: spec/presenters/post_presenter_spec.rb (hunk 16 above)
Comment (Rete2) [inline spec/presenters/post_presenter_spec.rb]: We did those changes and hope they are ok. Had to touch some other files now, but it was only because of the change of the method.

File: app/presenters/post_presenter.rb [hunk 17]
```diff
@@ -47,6 +47,10 @@ def page_title
     post_page_title @post
   end
 
+  def poll_participation_answer_id
```
Comment (SuperTux88) [inline app/presenters/post_presenter.rb]: The only reason you made this method public is because you need it in the tests. But you shouldn't change code only for tests. Can you please move this method down in the private scope (for example where `already_participated_in_poll` was).

File: spec/presenters/post_presenter_spec.rb [hunk 18]
```diff
@@ -135,6 +135,19 @@
       presenter = PostPresenter.new(status_message_with_poll)
       expect(presenter.as_json).to be_a(Hash)
     end
+
+    it "returns the answer id of the current user's poll participation" do
+      presenter = PostPresenter.new(status_message_with_poll, alice)
+      poll_answer = status_message_with_poll.poll.poll_answers.first
+      poll_participation = status_message_with_poll.poll.poll_participations
+      poll_participation = poll_participation.create(poll_answer: poll_answer, author: alice.person)
+      expect(presenter.poll_participation_answer_id).to eql(poll_participation.poll_answer_id)
```
Comment (SuperTux88) [inline spec/presenters/post_presenter_spec.rb]: You can replace this with:
```ruby
expect(presenter.to_json[:poll_participation_answer_id]).to eql(poll_participation.poll_answer_id)
```

File: spec/presenters/post_presenter_spec.rb [hunk 19]
```diff
@@ -135,6 +135,19 @@
...
+    end
+
+    it "returns nil if the user did not participate in a poll" do
+      presenter = PostPresenter.new(status_message_with_poll, alice)
+      expect(presenter.poll_participation_answer_id).to eql(nil)
```
Comment (SuperTux88) [inline spec/presenters/post_presenter_spec.rb]: Same as above.

**Review by SuperTux88 — CHANGES_REQUESTED** [review]
Looks good, only one small thing.

@svbergerem can you have a quick look at the frontend-change again (replace `already_participated_in_poll` with `poll_participation_answer_id`)? But I think it's OK.

File: spec/javascripts/jasmine_helpers/factory.js [hunk 20]
```diff
@@ -211,9 +211,9 @@ var factory = {
   },
 
   postWithPoll :  function(overrides) {
-    var defaultAttrs = _.extend(factory.postAttrs(),  {"author" : this.author()});
-    defaultAttrs = _.extend(defaultAttrs,  {"already_participated_in_poll" : false});
-    defaultAttrs = _.extend(defaultAttrs,  {"poll" : factory.poll()});
+    var defaultAttrs = _.extend(factory.postAttrs(), {"author": this.author()});
+    defaultAttrs = _.extend(defaultAttrs, {"poll_participation_answer_id": false});
```
Comment (svbergerem) [inline spec/javascripts/jasmine_helpers/factory.js]: This should be `null` instead of `false`, because that's what the server will return.

File: spec/javascripts/app/views/poll_view_spec.js [hunk 21]
```diff
@@ -65,7 +65,7 @@ describe("app.views.Poll", function(){
       expect(this.view.$('form').length).toBe(1);
     });
     it("hides vote form when user voted before", function(){
-      this.view.model.attributes.already_participated_in_poll = true;
+      this.view.model.attributes.poll_participation_answer_id = this.view.poll.poll_answers[0].id;
```
Comment (svbergerem) [inline spec/javascripts/app/views/poll_view_spec.js]: You can replace this by
```javascript
this.view.model.set("poll_participation_answer_id", this.view.poll.poll_answers[0].id);
```
to make pronto happy.

**Review by svbergerem — CHANGES_REQUESTED** [review]
Two minor remarks, everything else looks good.

File: spec/presenters/post_presenter_spec.rb (hunk 16 above)
Comment (netagonen) [inline spec/presenters/post_presenter_spec.rb]: And thanks for all the explanations!

**Comment by svbergerem** [comment]
I just found another small bug in the post preview while reviewing #7579:

poll preview

In the preview both the answer ids and the `poll_participation_answer_id` are undefined. We should only set `isCurrentUserVote` to `true` if the user gave an answer.

**Review by svbergerem — APPROVED** [review]
The front-end code looks good to me.

**Review by SuperTux88 — APPROVED** [review]
Backend looks good now.

**Comment by SuperTux88** [comment]
Merged as b556ad5211825cbc7fb711e6d0d3ca4e51666316

Thank you @Rete2 and @netagonen 🍪
